<?xml version="1.0"?>
<!--XML-->
    <launch>

    <!-- SETUP ARGUMENTS -->
    <!-- Be sure to change this to your cameras serials -->
    <arg name="camera_right_serial" default="31710260"/>
    <arg name="camera_left_serial" default="31710277"/>

    <arg name="camera_namespace" default="stereo"/>

    <arg name="frame_id" default="$(arg camera_namespace)_depth_optical_frame"/>

    <arg name="frame_width" default="2448"/>
    <arg name="frame_height" default="2048"/>
    <arg name="brightness" default="0"/>
    <arg name="exposure" default="10000"/>
    <arg name="gain" default="100"/>
    <arg name="exposure_auto" default="false"/>
    <arg name="gain_auto" default="false"/>
    <arg name="fps" default="5"/>
    <arg name="trigger" default="true"/>
    <!-- 0=TIS, 1=Basler -->
    <arg name="camera_type" default="0"/>
    <arg name="split_laser" default="false"/>

    <arg name="binning" default="1"/>

    <arg name="reverse_x" default="false"/>
    <arg name="reverse_y" default="false"/>

    <arg name="camera_info_right" default="file://${ROS_HOME}/camera_info/$(arg camera_namespace)/$(arg camera_namespace)_right.yaml" />
    <arg name="camera_info_left" default="file://${ROS_HOME}/camera_info/$(arg camera_namespace)/$(arg camera_namespace)_left.yaml" />

    <!-- STEREO <MATCHER> -->
    <group ns="$(arg camera_namespace)">

        <!-- Start the pylon control node for each camera -->
        <node if="$(eval camera_type == 1)" ns="left" name="i3dr_stereo_camera_left_ctrl" pkg="pylon_camera" type="pylon_camera_node" respawn="true" output="screen">
            <param name="camera_frame" value="$(arg frame_id)"/>
            <param name="device_user_id" value="$(arg camera_left_serial)"/>
            <param name="camera_info_url" value="$(arg camera_info_left)"/>
            <param name="frame_rate" value="$(arg fps)"/>
            <param name="exposure" value="$(arg exposure)"/>
            <param name="gain" value="$(arg gain)"/>
            <param name="brightness" value="$(arg brightness)"/>
            <param name="exposure_auto" value="$(arg exposure_auto)"/>
            <param name="gain_auto" value="$(arg gain_auto)"/>
            <param name="image_encoding" value="mono8"/>
            <param name="enable_hardware_trigger" value="true"/>
            <param name="mtu_size" value="3000"/>
            <param name="inter_pkg_delay" value="5000"/>
            <param name="enable_split_laser" value="$(arg split_laser)"/>
            <param name="binning_x" value="$(arg binning)"/>
            <param name="binning_y" value="$(arg binning)"/>
            <param name="reverse_x" value="$(arg reverse_x)"/>
            <param name="reverse_y" value="$(arg reverse_y)"/>
        </node>

        <node if="$(eval camera_type == 1)" ns="right" name="i3dr_stereo_camera_right_ctrl" pkg="pylon_camera" type="pylon_camera_node" respawn="true" output="screen">
            <param name="camera_frame" value="$(arg frame_id)"/>
            <param name="device_user_id" value="$(arg camera_right_serial)"/>
            <param name="camera_info_url" value="$(arg camera_info_right)"/>
            <param name="frame_rate" value="$(arg fps)"/>
            <param name="exposure" value="$(arg exposure)"/>
            <param name="gain" value="$(arg gain)"/>
            <param name="brightness" value="$(arg brightness)"/>
            <param name="exposure_auto" value="$(arg exposure_auto)"/>
            <param name="gain_auto" value="$(arg gain_auto)"/>
            <param name="image_encoding" value="mono8"/>
            <param name="enable_hardware_trigger" value="true"/>
            <param name="mtu_size" value="3000"/>
            <param name="inter_pkg_delay" value="0"/>
            <param name="enable_split_laser" value="$(arg split_laser)"/>
            <param name="binning_x" value="$(arg binning)"/>
            <param name="binning_y" value="$(arg binning)"/>
            <param name="reverse_x" value="$(arg reverse_x)"/>
            <param name="reverse_y" value="$(arg reverse_y)"/>
        </node>

        <!-- Start the tiscamera control node for each camera -->
        <!-- (optional) set inital camera properties using the parameters [Brightness/Exposure/Gain/Exposure_Auto/Gain_Auto] -->
        <!-- set camera properties using services from this node
             rosservice call /phobos/[left or right]/tiscam_[serial]_set_exposure_auto -->

        <node if="$(eval camera_type == 0)" ns="left" name="i3dr_stereo_camera_left_ctrl" pkg="i3dr_stereo_camera" type="tiscamera_ctrl.py" output="screen">
            <param name="camera_name" value="left"/>
            <param name="camera_info" value="$(arg camera_info_left)"/>
            <param name="camera_frame" value="$(arg frame_id)"/>
            <param name="Serial" value="$(arg camera_left_serial)"/>
            <param name="Width" value="$(arg frame_width)"/>
            <param name="Height" value="$(arg frame_height)"/>
            <param name="FPS" value="$(arg fps)"/>
            <param name="Brightness" value="$(arg brightness)"/>
            <param name="Exposure" value="$(arg exposure)"/>
            <param name="Gain" value="$(arg gain)"/>
            <param name="Exposure_Auto" value="$(arg exposure_auto)"/>
            <param name="Gain_Auto" value="$(arg gain_auto)"/>
            <param name="Trigger" value="$(arg trigger)"/>
            <param name="split_laser" value="$(arg split_laser)"/>
        </node>

        <node if="$(eval camera_type == 0)" ns="right" name="i3dr_stereo_camera_right_ctrl" pkg="i3dr_stereo_camera" type="tiscamera_ctrl.py" output="screen">
            <param name="camera_name" value="right"/>
            <param name="camera_info" value="$(arg camera_info_right)"/>
            <param name="camera_frame" value="$(arg frame_id)"/>
            <param name="Serial" value="$(arg camera_right_serial)"/>
            <param name="Width" value="$(arg frame_width)"/>
            <param name="Height" value="$(arg frame_height)"/>
            <param name="FPS" value="$(arg fps)"/>
            <param name="Brightness" value="$(arg brightness)"/>
            <param name="Exposure" value="$(arg exposure)"/>
            <param name="Gain" value="$(arg gain)"/>
            <param name="Exposure_Auto" value="$(arg exposure_auto)"/>
            <param name="Gain_Auto" value="$(arg gain_auto)"/>
            <param name="Trigger" value="$(arg trigger)"/>
            <param name="split_laser" value="$(arg split_laser)"/>
        </node>

        <node if="$(eval camera_type == 0)" name="i3dr_stereo_camera_trigger_ctrl" pkg="i3dr_stereo_camera" type="tiscamera_trigger.py" output="screen"/>

    </group>

</launch>
